library/core/src/num/nonzero.rs

```diff
@@ -83,6 +83,90 @@ impl_zeroable_primitive!(NonZeroIsize(isize));
 #[rustc_diagnostic_item = "NonZero"]
 pub struct NonZero<T: ZeroablePrimitive>(T);
 
+impl<T> NonZero<T>
+where
+    T: ZeroablePrimitive,
+{
+    /// Creates a non-zero if the given value is not zero.
+    #[stable(feature = "nonzero", since = "1.28.0")]
+    #[rustc_const_stable(feature = "const_nonzero_int_methods", since = "1.47.0")]
+    #[rustc_allow_const_fn_unstable(const_refs_to_cell)]
```

Comment:
Uh, is this needed?

Reply:
for the `transmute_copy` taking a reference to a generic parameter. Unfortunately we can't use `transmute`, even though we know the size is the same, rustc doesn't.

We could teach the transmute checks about `rustc_nonnull_optimization_guaranteed` and use `transmute` again.